app/controllers/users.rb

```diff
@@ -0,0 +1,36 @@
+# get a list of users
+get '/users' do
+end
+# get a form to add a new users
+get '/users/new' do
+
+  erb :new_user
+end
+# add a new users
+post '/users' do
+  session[:errors] = nil
+  user = User.create(username: params[:username], email: params[:email], password: params[:password])
+  if user.errors.any?
+    session[:errors] = user.display_errors
+    redirect '/users/new' #<< partial (: invalid_user )
```

Comment:
instead of putting errors in the session and redirecting you should just render right now with the errors in an instance variable. Like this
```ruby
post '/users' do
  @user = User.new(username: params[:username], email: params[:email], password: params[:password])
  if @user.save
    session[:user_id] = @user.id
    redirect '/'
  else
    erb :new_user
  end
end
```

The `new_user` should use the `@user` instance variable.